Applications/Spire/Source/TimeAndSales/TimeAndSalesTableModel.cpp

```diff
@@ -95,31 +98,17 @@ void TimeAndSalesTableModel::load_snapshot(Queries::Sequence last, int count) {
   m_promise = m_model->query_until(last, count).then(
     [=] (auto&& result) {
       auto& snapshot = result.Get();
-      for(auto i = snapshot.rbegin(); i != snapshot.rend(); ++i) {
-        m_entries.push(*i);
-      }
+      m_transaction.transact([&] {
+        for(auto i = snapshot.rbegin(); i != snapshot.rend(); ++i) {
+          m_entries.push_back(*i);
+          m_transaction.push(TableModel::AddOperation(m_entries.size() - 1));
+        }
+      });
       m_end_loading_signal();
     });
 }
 
 void TimeAndSalesTableModel::on_update(const TimeAndSalesModel::Entry& entry) {
-  m_entries.insert(entry, 0);
-}
-
-void TimeAndSalesTableModel::on_operation(
-    const ListModel<TimeAndSalesModel::Entry>::Operation& operation) {
-  visit(operation,
-    [&] (const StartTransaction&) {
-      m_transaction.start();
-    },
-    [&] (const EndTransaction&) {
-      m_transaction.end();
-    },
-    [&] (const ListModel<TimeAndSalesModel::Entry>::AddOperation& operation) {
-      m_transaction.push(TableModel::AddOperation(operation.m_index));
-    },
-    [&] (const ListModel<TimeAndSalesModel::Entry>::RemoveOperation&
-        operation) {
-      m_transaction.push(TableModel::RemoveOperation(operation.m_index));
-    });
+  m_entries.push_front(entry);
```

Comment:
```std::deque``` is appropriate when you need a double ended queue, that is you're pushing and popping from the front and back. It's not suitable for this task where only insertions to one end are performed. A dynamic array is the correct data structure for a list of elements, to preserve the ```O(1)``` guarantees ensure you only push and pop to the end of the array.

Reply:
When the model queries historical data, data is pushed to the back. When the model gets real-time data, data is inserted to the front.  Hence, I think this task requires inserting data into the front and back ends, not just one end.

Reply:
Querying and inserting historical data is already an ```O(N)``` operation, there is nothing that can be done to reduce it's time complexity.

Reply:
I am confused. For this task, data can only be inserted to the front or pushed to the back, but not to the middle. `std::deuqe` has constant O(1) for inserting elements at the end or beginning, so I think it is suitable for this case.

Reply:
Come up with a solution that uses a dynamic array instead of a std::deque. A deque is not the appropriate data structure for this use case, it's too slow.